File: location/src/test/java/org/odk/collect/location/GoogleFusedLocationClientTest.java
package org.odk.collect.location;

import static android.location.LocationManager.GPS_PROVIDER;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;
import static org.junit.Assert.assertSame;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doAnswer;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.reset;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import android.location.Location;
import android.location.LocationManager;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.location.FusedLocationProviderApi;
import com.google.android.gms.location.LocationListener;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.invocation.InvocationOnMock;
import org.mockito.stubbing.Answer;
import org.odk.collect.testshared.LocationTestUtils;

@RunWith(AndroidJUnit4.class)
public class GoogleFusedLocationClientTest {

    private GoogleApiClient googleApiClient;
    private GoogleFusedLocationClient client;
    private FusedLocationProviderApi fusedLocationProviderApi;

    @Before
    public void setUp() {
        fusedLocationProviderApi = mock(FusedLocationProviderApi.class);
        googleApiClient = mock(GoogleApiClient.class);
        LocationManager locationManager = mock(LocationManager.class);
        client = new GoogleFusedLocationClient(googleApiClient, fusedLocationProviderApi, locationManager);
    }

    @Test
    public void startShouldCallLocationClientOnConnected() {
        doAnswer(new OnConnectedAnswer()).when(googleApiClient).connect();

        LocationClient.LocationClientListener listener = mock(LocationClient.LocationClientListener.class);
        client.setListener(listener);

        client.start();
        verify(listener).onClientStart();
        verify(listener, never()).onClientStartFailure();
        verify(listener, never()).onClientStop();

        client.stop();

        reset(listener);
        doAnswer(new OnConnectionFailedAnswer()).when(googleApiClient).connect();

        client.start();
        verify(listener, never()).onClientStart();
        verify(listener).onClientStartFailure();
        verify(listener, never()).onClientStop();

        client.stop();
        verify(listener, never()).onClientStart();
        verify(listener).onClientStartFailure();
        verify(listener).onClientStop();
    }

    @Test
    public void stopShouldDisconnectFromGoogleApiIfConnected_andAlwaysCallOnClientStopIfListenerSet() {
        LocationClient.LocationClientListener listener = mock(LocationClient.LocationClientListener.class);
        client.setListener(listener);

        // Previously connected, disconnection succeeds
        when(googleApiClient.isConnected()).thenReturn(true);
        client.stop();
        verify(googleApiClient).disconnect();
        verify(listener).onClientStop();

        reset(listener);

        // Previously connected, disconnection calls onConnectionSuspended
        doAnswer(new OnDisconnectedAnswer()).when(googleApiClient).disconnect();
        client.stop();
        verify(listener).onClientStop();

        reset(listener);

        // Not previously connected
        when(googleApiClient.isConnected()).thenReturn(false);
        client.stop();
        verify(listener).onClientStop();
    }

    @Test
    public void whenGoogleApiClientNotConnected_shouldNotRemoveLocationUpdatesBeCalled() {
        when(googleApiClient.isConnected()).thenReturn(false);

        client.start();

        TestLocationListener listener = new TestLocationListener();
        client.requestLocationUpdates(listener);

        client.stop();
        verify(fusedLocationProviderApi, never()).removeLocationUpdates(googleApiClient, listener);
    }

    @Test
    public void whenGoogleApiClientNotConnected_shouldNotRequestLocationUpdatesBeCalled() {
        when(googleApiClient.isConnected()).thenReturn(false);

        client.start();
        client.requestLocationUpdates(new TestLocationListener());

        verify(fusedLocationProviderApi, never()).requestLocationUpdates(any(), any(), (LocationListener) any());
    }

    @Test
    public void requestingLocationUpdatesShouldUpdateCorrectListener() {
        client.start();

        TestLocationListener firstListener = new TestLocationListener();
        client.requestLocationUpdates(firstListener);

        Location firstLocation = newMockLocation();
        client.onLocationChanged(firstLocation);

        assertSame(firstLocation, firstListener.getLastLocation());

        Location secondLocation = newMockLocation();
        client.onLocationChanged(secondLocation);

        assertSame(secondLocation, firstListener.getLastLocation());

        // Now stop updates:
        client.stopLocationUpdates();

        Location thirdLocation = newMockLocation();
        client.onLocationChanged(thirdLocation);

        // Should still be second:
        assertSame(secondLocation, firstListener.getLastLocation());

        // Call requestLocationUpdates again with new listener:
        TestLocationListener secondListener = new TestLocationListener();
        client.requestLocationUpdates(secondListener);

        Location fourthLocation = newMockLocation();
        client.onLocationChanged(fourthLocation);

        // First listener should still have second location:
        assertSame(secondLocation, firstListener.getLastLocation());
        assertSame(fourthLocation, secondListener.getLastLocation());

        // Call stop() and make sure it called stopLocationUpdates():
        client.stop();

        Location fifthLocation = newMockLocation();
        client.onLocationChanged(fifthLocation);

        // Listener should still have fourth location:
        assertSame(fourthLocation, secondListener.getLastLocation());
    }

    @Test
    public void getLastLocationShouldCallBlockingConnectIfNotConnected() {
        client.getLastLocation();
        verify(googleApiClient).blockingConnect();

        when(googleApiClient.isConnected()).thenReturn(true);
        client.start();

        client.getLastLocation();
        verify(googleApiClient).blockingConnect(); // 'verify' checks if called *once*.
    }

    @Test
    public void whenNewlyReceivedLocationAccuracyIsNegative_shouldBeSetToZero() {
        TestLocationListener listener = new TestLocationListener();
        client.requestLocationUpdates(listener);

        Location location = LocationTestUtils.createLocation("GPS", 7, 2, 3, -1);
        client.onLocationChanged(location);

        assertThat(listener.getLastLocation().getAccuracy(), is(0.0f));
    }

    @Test
    public void whenNewlyReceivedLocationIsMocked_shouldAccuracyBeSetToZero() {
        TestLocationListener listener = new TestLocationListener();
        client.requestLocationUpdates(listener);

        Location location = LocationTestUtils.createLocation("GPS", 7, 2, 3, 5.0f, true);
        client.onLocationChanged(location);

        assertThat(listener.getLastLocation().getAccuracy(), is(0.0f));
    }

    @Test
    public void whenNewlyReceivedLocationIsMocked_andRetainMockAccuracyIsTrue_doesNotChangeAccuracy() {
        TestLocationListener listener = new TestLocationListener();
        client.setRetainMockAccuracy(true);
        client.requestLocationUpdates(listener);

        Location location = LocationTestUtils.createLocation("GPS", 7, 2, 3, 5.0f, true);
        client.onLocationChanged(location);

        assertThat(listener.getLastLocation().getAccuracy(), is(5.0f));
    }

    @Test
    public void whenLastKnownLocationAccuracyIsNegative_shouldBeSetToZero() {
        Location location = LocationTestUtils.createLocation(GPS_PROVIDER, 7, 2, 3, -1.0f);
        when(fusedLocationProviderApi.getLastLocation(googleApiClient)).thenReturn(location);

        assertThat(client.getLastLocation().getAccuracy(), is(0.0f));
    }

    @Test
    public void whenLastKnownLocationIsMocked_shouldAccuracyBeSetToZero() {
        Location location = LocationTestUtils.createLocation(GPS_PROVIDER, 7, 2, 3, 5.0f, true);
        when(fusedLocationProviderApi.getLastLocation(googleApiClient)).thenReturn(location);

        assertThat(client.getLastLocation().getAccuracy(), is(0.0f));
    }

    @Test
    public void whenLastKnownLocationIsMocked_andRetainMockAccuracyIsTrue_doesNotChangeAccuracy() {
        Location location = LocationTestUtils.createLocation(GPS_PROVIDER, 7, 2, 3, 5.0f, true);
        client.setRetainMockAccuracy(true);
        when(fusedLocationProviderApi.getLastLocation(googleApiClient)).thenReturn(location);

        assertThat(client.getLastLocation().getAccuracy(), is(5.0f));
    }

    private static Location newMockLocation() {
        return mock(Location.class);
    }

    private class OnConnectedAnswer implements Answer<Object> {
        @Override
        public Object answer(InvocationOnMock invocation) throws Throwable {
            client.onConnected(null);
            return null;
        }
    }

    private class OnConnectionFailedAnswer implements Answer<Object> {
        @Override
        public Object answer(InvocationOnMock invocation) throws Throwable {
            client.onConnectionFailed(new ConnectionResult(0));
            return null;
        }
    }

    private class OnDisconnectedAnswer implements Answer<Object> {
        @Override
        public Object answer(InvocationOnMock invocation) throws Throwable {
            client.onConnectionSuspended(0);
            return null;
        }
    }
}

File: location/src/test/java/org/odk/collect/location/GoogleFusedLocationClientTest.kt
package org.odk.collect.location

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.google.android.gms.location.LocationRequest
import org.hamcrest.CoreMatchers.equalTo
import org.hamcrest.CoreMatchers.notNullValue
import org.hamcrest.CoreMatchers.nullValue
import org.hamcrest.MatcherAssert.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.Mockito.verify
import org.mockito.kotlin.any
import org.mockito.kotlin.argThat
import org.mockito.kotlin.isA
import org.mockito.kotlin.mock
import org.odk.collect.location.LocationClient.LocationClientListener
import org.odk.collect.testshared.LocationTestUtils.createLocation

@RunWith(AndroidJUnit4::class)
class GoogleFusedLocationClientTest {
    private val fusedLocationProviderClientWrapper = mock<FusedLocationProviderClientWrapper>()
    private val locationClientListener = mock<LocationClientListener>()
    private val locationListener = TestLocationListener()
    private lateinit var client: GoogleFusedLocationClient

    @Before
    fun setUp() {
        client = GoogleFusedLocationClient(fusedLocationProviderClientWrapper, mock())
    }

    @Test
    fun `#start sets listener and calls #onClientStart`() {
        client.start(locationClientListener)

        assertThat(client.getListener(), notNullValue())
        verify(locationClientListener).onClientStart()
    }

    @Test
    fun `#start starts FusedLocationProviderClientWrapper`() {
        client.start(locationClientListener)

        verify(fusedLocationProviderClientWrapper).start(isA<GoogleFusedLocationClient>())
    }

    @Test
    fun `#stop removes listener and calls #onClientStop`() {
        client.start(locationClientListener)
        client.stop()

        assertThat(client.getListener(), nullValue())
        verify(locationClientListener).onClientStop()
    }

    @Test
    fun `#stop stops location updates`() {
        val listener = TestLocationListener()
        client.requestLocationUpdates(listener)
        client.stop()

        verify(fusedLocationProviderClientWrapper).removeLocationUpdates()
    }

    @Test
    fun `#requestLocationUpdates starts location updates`() {
        val listener = TestLocationListener()
        client.requestLocationUpdates(listener)

        verify(fusedLocationProviderClientWrapper).requestLocationUpdates(any())
    }

    @Test
    fun `#requestLocationUpdates builds location request based on passed values`() {
        client.setPriority(LocationClient.Priority.PRIORITY_NO_POWER)
        client.setUpdateIntervals(10L, 5L)
        val listener = TestLocationListener()
        client.requestLocationUpdates(listener)

        verify(fusedLocationProviderClientWrapper).requestLocationUpdates(
            argThat { locationRequest: LocationRequest ->
                assertThat(locationRequest.priority, equalTo(LocationClient.Priority.PRIORITY_NO_POWER.value))
                assertThat(locationRequest.interval, equalTo(10L))
                assertThat(locationRequest.fastestInterval, equalTo(5L))
                true
            }
        )
    }

    @Test
    fun `#isMonitoringLocation returns false if listener is mot set`() {
        client.stopLocationUpdates()

        assertThat(client.isMonitoringLocation, equalTo(false))
    }

    @Test
    fun `#isMonitoringLocation returns true if listener is set`() {
        val listener = TestLocationListener()
        client.requestLocationUpdates(listener)

        assertThat(client.isMonitoringLocation, equalTo(true))
    }

    @Test
    fun `#stopLocationUpdates stops location updates`() {
        val listener = TestLocationListener()
        client.requestLocationUpdates(listener)
        client.stopLocationUpdates()

        verify(fusedLocationProviderClientWrapper).removeLocationUpdates()
    }

    @Test
    fun `when received location accuracy is negative set to zero`() {
        client.requestLocationUpdates(locationListener)
        val location = createLocation("GPS", 7.0, 2.0, 3.0, -1f)
        client.onLocationChanged(location)

        assertThat(locationListener.lastLocation!!.accuracy, equalTo(0.0f))
    }

    @Test
    fun `when received location accuracy is mocked set to zero`() {
        client.requestLocationUpdates(locationListener)
        val location = createLocation("GPS", 7.0, 2.0, 3.0, 5.0f, true)
        client.onLocationChanged(location)

        assertThat(locationListener.lastLocation!!.accuracy, equalTo(0.0f))
    }

    @Test
    fun `when received location accuracy is mocked and #retainMockAccuracy is true do not change accuracy`() {
        client.setRetainMockAccuracy(true)
        client.requestLocationUpdates(locationListener)
        val location = createLocation("GPS", 7.0, 2.0, 3.0, 5.0f, true)
        client.onLocationChanged(location)

        assertThat(locationListener.lastLocation!!.accuracy, equalTo(5.0f))
    }
}
